renormalizer/mps/lib.py

```diff
@@ -97,6 +116,54 @@ def read(self, domain: str, siteidx: int):
         return asxp(self._virtual_disk[(domain, siteidx)])
 
 
+def contract_one_site_multi_mpo(environ, ms, mos, domain, ms_conj=None):
```

Comment:
The function name and documentation implies contraction over more than 2 MPOs is also supported. 
I don't know if the circumstance will appear in the future, but I think it is possible to write a generic routine based on `opt_einsum`. Another benefit is that `contract_one_site_multi_mpo` and `contract_one_site` can be united.

This is not urgent and can be postponed to future works.

Reply:
In `test_mp.py`, I add a test `test_environ_multi_mpo` to make sure `contract_one_site_multi_mpo` could generalize `contract_one_site`.  As you said `cv` could use it to construct the environment. In this commit, I use this function to optimize the ground state of (H-\omega)^2. 
I can do this maintenance in the future.

Reply:
> In this commit, I use this function to optimize the ground state of (H-\omega)^2.

Speaking of which, is it feasible to construct `(H-\omega)^2` as an MPO? *Maybe* the resulting scaling is lower.

Reply:
You mean we manually expand (H-omega)^2 as a single operator (eliminate and combine some terms are possible) or just multiply two MPO (H-\omega)? If the latter, the scaling is higher (think about constructing the environment) and meanwhile the memory usage is also higher.

Reply:
I mean the first one, taking advantage of the general MPO construction routine and `Op` multiplication operations. Maybe we can let an undergraduate student test it :)

Reply:
It is possible that the scaling would be smaller in some cases (maybe in the Holstein-like model). Then, it could be used to reduce the cost of the `cv` algorithm. 
However, in QC (or vibrational Hamiltonian, or other general long-range Hamiltonians too), the H^2 has MPO with O(N^4) just as the direct multiplication of two H (O(N^2) \times O(N^2)).  In these cases, the scaling is worse.